frappe/integrations/doctype/connected_app/connected_app.py

```diff
@@ -102,17 +102,39 @@ def get_scopes(self):
 	def get_query_params(self):
 		return {param.key: param.value for param in self.query_parameters}
 
+	def get_active_token(self, user=None):
+		user = user or frappe.session.user
+		token_cache = self.get_token_cache(user)
+		if token_cache and token_cache.is_expired():
+			oauth_session = self.get_oauth2_session(user)
+
+			try:
+				token = oauth_session.refresh_token(
+					body=f"redirect_uri={self.redirect_uri}",
+					token_url=self.token_uri,
+					refresh_token=token_cache.get_password("refresh_token"),
+				)
+			except Exception as e:
+				msg = _("An error occurred while refreshing the access token for Connected App {0}: ").format(
+					self.name
+				)
+				title = "Oauth Token Refresh Error"
+				frappe.throw(msg + str(e), title=_(title))
+				self.log_error(title, str(e))
```

Comment:
currently only added logging in the case any exception occurs while refreshing the access token.

i'm kinda skeptical about the expiry of refresh token as the user will have to again authorize to get a new refresh token - any thoughts on this are welcome.

ref (for google): https://developers.google.com/identity/protocols/oauth2#expiration - they've not stated anything for prod usage so i'm kinda guessing the token will remain active for a long time (unless the user does someting)

Reply:
silent logging is fine. emails are also sent in background from workers

Whatever the refresh limits are the will be followed. If there is no refresh limit (generally in case of production, verified by provider clients) the emails work forever, unless user goes to google's connected apps section and removes/revokes the client.